### ISSUE TITLE
Softmax() is not working on GPU

### ISSUE THREAD

**Issue by sensus-sextus** [NervanaSystems/neon#112]
Hello,
I have slightly modified the "mnist_mlp.py" example from Neon's distribution.
I'm using softmax as an activation function of the output layer instead of original logistic.
Also I added fprop to get trained network's predictions on the MNIST standard test set

This works well on CPU, and I obtained ~95% classification accuracy on this test.
But the same example goes wrong on GPU here only 1 example out of 1000 was classified correctly.
Meantime,  after training the network reported only 4.53% of misclassification!

We consider this behavior as  a bug, thank you in advance for fixing it.
Best wishes.

**Comment by ursk** [comment]
Hi,
by your own words, you get 95% _accuracy_ on CPU and 5% _misclassification_ on GPU, so it doesn’t seem like a bug.  We checked the code and did not find any issues.

**Comment by sensus-sextus** [comment]
Sorry for insisting, but  the main isssue is that such a network cannot predict the class when trained on GPU!!!!

The actual accuracy (what I measured using fprop()) was 95% and 0.1% on CPU and GPU respectively, so misclassification rate was 5% and 99.9%
On the contrary, if checked by mlp.eval(valid_set, metric=Misclassification()), the reported rate was close to 5% in both cases.

**Comment by apark263** [comment]
if you're using softmax as an output activation, your cost function has to be multiclass cross entropy, not binomial cross entropy.  we don't support softmax -> binomial cross entropy or logistic -> multiclass cross entropy.

**Comment by sensus-sextus** [comment]
In all this experiments I'm using SumSquared() as a cost.
It works perfectly on CPU but fails on GPU.
I think this error is something technical not algorithmic.

**Comment by ursk** [comment]
It's a bit hard for us to follow what is required to reproduce the behavior that you are seeing. For example if you are training with a SumSquared cost, it seems like the model is doing regression, and it's not clear how you measure misclassification. 

Generally, if you find a bug, the best way to go is to provide a minimal example that reproduces it. Preferably this would be modifying one of the examples slightly, but still using the original example data so that we can run it. We will then do our best to provide a speedy fix.

**Comment by sensus-sextus** [comment]
Actually, this situation does not depend on the last-layer activation function (softmax or logistic) 
I was able to fix it by using the following code to get predictions:
X0= X_test [0:1000,:]
#X1= X0.transpose()
xx= gp.array(X0, dtype=float32)
xx1=xx.T
prdc= mlp.fprop(xx1)

For some reasons that are unknown to me, if I transpose the array before putting it to backend, the prediction fails, as I described above
